### DISCUSSION TITLE
Validation on submit does not consider manually set errors

### DISCUSSION THREAD

**Discussion by dasMoeppi** [react-hook-form/react-hook-form#1906] [Q&A] [answered]
Hi,

first of all, I really love this forms library, thank you for your work!
But right now I am facing a problem I would consider a flaw.
I noticed that every time the `handleSubmit` function is called by clicking on a submitButton, the validation is triggered as well. Normally, this is not much of a problem. But I make use of the `setError` function to add errors to the form manually. The current implementation "swallows" my custom errors since they are set outside of the `validate` function.
I have created a sandbox to show you what I mean: 
https://codesandbox.io/s/rhf-submit-validation-0udwz?file=/src/ExampleForm.js

1. Add a value to the input field
2. Add an error manually by clicking the button
3. Submit

As you can see, the form will be valid after hitting the submit-button because the `validate` function checks for the required status only of course.
I think it is a bit weird that the validation is triggered although the form already has an invalid status. Normally, I would expect that an invalid form can't be submitted at all.

Can you help me? Is this intended or rather a bug?

Kind regards

**Comment by bluebill1049** [comment, marked as answer]
Thanks for your feedback @mjbertenburg This is the same issue https://github.com/react-hook-form/react-hook-form/issues/1881 please have a read there.

**Reply by dasMoeppi** [reply]
Hi,
wow, thank you for your fast reply! You are right, I am struggling with the same issue. I should have read the documentation about `setError`more carefully, my bad.
Unfortunately, the proposed solutions will not work in my case. Is there any way to prevent the form from revalidating on submitting?
Maybe by wrapping RHF's `handleSubmit` function into another function which checks the valid-state of the form and prevents the handleSubmit actions?

**Reply by bluebill1049** [reply]
maybe should have some sort config for persist error which requires users to manually clear 🤔

**Reply by dasMoeppi** [reply]
Probably. Right now I am doing the cleanup for these custom errors manually.
At least for my use case, either a config option for persisting manual errors or to prevent submitting an invalid form would be helpful. 
The decision is yours of course. Have other people been struggling with the non-persisting manual errors besides https://github.com/react-hook-form/react-hook-form/issues/1881 ?

**Reply by bluebill1049** [reply]
Thanks i understand the concern, i think his problem and yours are different after careful thought.

1. you want to prevent submit
2. he wants to persist the error not prevent submit

i think this needs some careful thought before implementation, otherwise, it would be a very confusing topic.

**Reply by bluebill1049** [reply]
I am addressing this issue: https://github.com/react-hook-form/react-hook-form/pull/1907 let me know your thoughts.

**Reply by dasMoeppi** [reply]
>Thanks i understand the concern, i think his problem and yours are different after careful thought. 
>1. you want to prevent submit 
>2. he wants to persist the error not prevent submit

You are right, these are two different problems. As far as I understand the merged bugfix, errors set via `setError` will now be persisted, so problem no. 2 should be solved now.
I guess, I need to find an own workaround for my use case for now. The reason why I would like to prevent validation on submitting if the form is invalid, is that one of my fields has an asynchronous validator. So, every time you hit submit on the invalid form an unnecessary API call is fired. ;)
